### PR TITLE
types: add interactions types

### PR DESCRIPTION
**Pull request by itohatweb**

### CODE REVIEW

**Comment by ayntee** [comment]
Rename the sub-module to commands to follow convention (see `handlers` and `helpers`)

File: src/types/interactions/application_commands/application_command_callback_data.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  tts?: boolean;
+  /** Message Content */
+  content?: string;
+  /** Supports up to 10 embeds */
+  embeds?: Embed[];
+  /** Allowed Mentions object */
+  allowedMentions?: AllowedMentions;
+  /** Set to `64` to make your response ephemeral */
+  flags?: number;
+}
```
Comment (ayntee) [inline src/types/interactions/application_commands/application_command_callback_data.ts:16]: Screenshot_20210329_162137
Is this correct?

**Comment by itohatweb** [comment]
>Rename the sub-module to commands to follow convention (see handlers and helpers). Also, I would much prefer if there is no nested folder inside of the commands sub-module (now interactions).

commands are not interactions.
Discord will add different interaction types later so a separate `commands` folder would make sense.

File: src/types/interactions/application_commands/application_command_callback_data.ts (hunk 1 above)
Comment (itohatweb) [reply src/types/interactions/application_commands/application_command_callback_data.ts:16]: wrong interface name

Comment (ayntee) [reply src/types/interactions/application_commands/application_command_callback_data.ts:16]: See the file name :roll_eyes:

File: src/types/interactions/application_commands/application_command_interaction_data.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  /** The name of the invoked command */
+  name: string;
+  /** The params + values from the user */
+  options?: ApplicationCommandInteractionDataOption[];
+}
+
+/** https://discord.com/developers/docs/interactions/slash-commands#interaction-applicationcommandinteractiondata */
+export type DiscordApplicationCommandInteractionData = SnakeCaseProps<
+  ApplicationCommandInteractionData
+>;
```
Comment (ayntee) [inline src/types/interactions/application_commands/application_command_interaction_data.ts:15]: Duplicate, either change the previously reviewed change or you may change this.

**Comment by ayntee** [comment]
> > Rename the sub-module to commands to follow convention (see handlers and helpers). Also, I would much prefer if there is no nested folder inside of the commands sub-module (now interactions).
> 
> commands are not interactions.
> Discord will add different interaction types later so a separate `commands` folder would make sense.

Hmm, you are somewhat right, but can you un-nest the files and move them under `interactions/*` to just `interactions/`?

**Comment by ayntee** [comment]
I will update helpers and handlers module accordingly.

File: src/types/interactions/application_commands/application_command_callback_data.ts (hunk 1 above)
Comment (itohatweb) [reply src/types/interactions/application_commands/application_command_callback_data.ts]: Somehow my git didn't push right

File: src/types/interactions/commands/command.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { SnakeCaseProps } from "../../util.ts";
+import { CommandOption } from "./command_option.ts";
+
+export interface Command {
```
Comment (ayntee) [inline src/types/interactions/commands/command.ts:4]: This should be `ApplicationCommand`, not `Command`

File: src/types/interactions/application_command_callback_data.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { Embed } from "../../embeds/embed.ts";
+import { AllowedMentions } from "../../messages/allowed_mentions.ts";
+import { SnakeCaseProps } from "../../util.ts";
+
+export interface ApplicationCommandCallbackData {
```
Comment (ayntee) [inline src/types/interactions/application_command_callback_data.ts:5]: Rename to `InteractionApplicationCommandCallbackData`
```suggestion
export interface InteractionApplicationCommandCallbackData {
```

File: src/types/interactions/application_command_callback_data.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  embeds?: Embed[];
+  /** Allowed Mentions object */
+  allowedMentions?: AllowedMentions;
+  /** Set to `64` to make your response ephemeral */
+  flags?: number;
+}
+
+/** https://discord.com/developers/docs/interactions/slash-commands#interaction-response-interactionapplicationcommandcallbackdata */
+export type DiscordApplicationCommandCallbackData = SnakeCaseProps<
+  ApplicationCommandCallbackData
```
Comment (ayntee) [inline src/types/interactions/application_command_callback_data.ts:20]:
```suggestion
export type DiscordInteractionApplicationCommandCallbackData = SnakeCaseProps<
  InteractionApplicationCommandCallbackData
```

File: src/types/interactions/application_command_callback_data.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { Embed } from "../../embeds/embed.ts";
+import { AllowedMentions } from "../../messages/allowed_mentions.ts";
+import { SnakeCaseProps } from "../../util.ts";
+
+export interface ApplicationCommandCallbackData {
+  /** Is the response TTS */
+  tts?: boolean;
+  /** Message Content */
```
Comment (ayntee) [inline src/types/interactions/application_command_callback_data.ts:8]:
```suggestion
  /** Message content */
```

File: src/types/interactions/application_command_option.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import { SnakeCaseProps } from "../../util.ts";
+import { ApplicationCommandOptionChoice } from "./command_option_choice.ts";
+
+export interface ApplicationCommandOption {
+  /** Value of Application Command Option Type */
+  type: number;
```
Comment (ayntee) [inline src/types/interactions/application_command_option.ts:6]: We should have a separate enum and cast that enum to the type of property "type".

File: src/types/interactions/interaction.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { User } from "../users/user.ts";
+import { SnakeCaseProps } from "../util.ts";
+import { ApplicationCommandCallbackData } from "./application_command_callback_data.ts";
+import { InteractionTypes } from "./interaction_types.ts";
+
+export interface Interaction {
+  /** Id of the interaction */
+  id: string;
+  /** The type of interaction */
+  type: InteractionTypes;
```
Comment (ayntee) [inline src/types/interactions/interaction.ts:10]: Rename `InteractionTypes` to `DiscordInteractionTypes`

File: src/types/interactions/interaction.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import { User } from "../users/user.ts";
+import { SnakeCaseProps } from "../util.ts";
+import { ApplicationCommandCallbackData } from "./application_command_callback_data.ts";
+import { InteractionTypes } from "./interaction_types.ts";
+
+export interface Interaction {
+  /** Id of the interaction */
```
Comment (ayntee) [inline src/types/interactions/interaction.ts:7]:
```suggestion
  /** id of the interaction */
```

**Review by ayntee — APPROVED** [review]

**Pull request merged**